src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java

```diff
@@ -15,15 +15,182 @@
+
 import java.io.Closeable;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
 import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.concurrent.TimeUnit;
 
 /**
  * An iterator which iterates over every element on all the pages of the given resource. The page urls are generated
  * based on configured {@link PaginationType}.
  */
 public abstract class BaseHttpPaginationIterator implements Iterator<PageEntry>, Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(BaseHttpPaginationIterator.class);
+
+  protected final BaseHttpSourceConfig config;
+  private final HttpClient httpClient;
+  private final HttpErrorHandler httpErrorHandler;
+
+  protected String nextPageUrl;
+  private BasePage page;
+  private Integer httpStatusCode;
+  private CloseableHttpResponse response;
+
+  public BaseHttpPaginationIterator(BaseHttpSourceConfig config) {
+    this.config = config;
+    this.httpClient = new HttpClient(config);
+    this.nextPageUrl = config.getUrl();
+    this.httpErrorHandler = new HttpErrorHandler(config);
+  }
+
+  protected abstract String getNextPageUrl(String body, CloseableHttpResponse response, BasePage page);
   public abstract boolean supportsSkippingPages();
 
-  // TODO: this class is not yet implemented!
+  protected boolean visitPageAndCheckStatusCode() throws IOException {
+    if (response != null) { // close previous response
+      response.close();
+    }
+
+    response = httpClient.executeHTTP(nextPageUrl);
+    httpStatusCode = response.getStatusLine().getStatusCode();
+    HttpErrorHandlingStrategy errorHandlingStrategy = httpErrorHandler.getErrorHandlingStrategy(httpStatusCode);
+
+    return !errorHandlingStrategy.shouldRetry();
+  }
+
+  protected BasePage getNextPage() throws IOException, InterruptedException {
+    // no more pages
+    if (nextPageUrl == null) {
+      return null;
+    }
+
+    // response being null, means it's the first page we are loading
+    if (response != null && config.getWaitTimeBetweenPages() != null) {
+      Thread.sleep(config.getWaitTimeBetweenPages());
+    }
+
+    LOG.debug("Fetching '{}'", nextPageUrl);
+
+    PollInterval pollInterval;
+    if (config.getRetryPolicy().equals(RetryPolicy.LINEAR)) {
+      pollInterval = FixedPollInterval.fixed(config.getLinearRetryInterval(), TimeUnit.SECONDS);
+    } else {
+      pollInterval = IterativePollInterval.iterative(duration -> duration.multiply(2));
+    }
+
+    try {
+      Awaitility
+        .await().with()
+        .pollInterval(pollInterval)
+        .timeout(config.getMaxRetryDuration(), TimeUnit.SECONDS)
+        .until(this::visitPageAndCheckStatusCode);
+    } catch (ConditionTimeoutException ex) {
+      // Retries failed. We don't need to do anything here. This will ba handled using httpStatusCode below.
+    }
+
+    HttpEntity reponseEntity = response.getEntity();
+    byte[] responseBytes = EntityUtils.toByteArray(reponseEntity);
```

Comment:
At some point we will need to read whole byte array into memory. Since it needs to be written to structured record "body" field. Am I missing something?

Reply:
That's only for the 'blob' format. Everything else should not. You can ignore this now for now, since format has been moved out.